team_project_access.go

```diff
@@ -57,14 +58,106 @@ type TeamProjectAccessList struct {
 
 // TeamProjectAccess represents a project access for a team
 type TeamProjectAccess struct {
-	ID     string                `jsonapi:"primary,team-projects"`
-	Access TeamProjectAccessType `jsonapi:"attr,access"`
+	ID              string                      `jsonapi:"primary,team-projects"`
+	Access          TeamProjectAccessType       `jsonapi:"attr,access"`
+	ProjectAccess   *ProjectAccessPermissions   `jsonapi:"attr,project-access"`
+	WorkspaceAccess *WorkspaceAccessPermissions `jsonapi:"attr,workspace-access"`
 
 	// Relations
 	Team    *Team    `jsonapi:"relation,team"`
 	Project *Project `jsonapi:"relation,project"`
 }
 
+// ProjectAccessPermissions represents the team's specific permissions related to the project itself
+type ProjectAccessPermissions struct {
+	ProjectSettingsPermission ProjectSettingsPermissionType `jsonapi:"attr,settings"`
+	ProjectTeamsPermission    ProjectTeamsPermissionType    `jsonapi:"attr,teams"`
+}
+
+// WorkspaceAccessPermissions represents the team's specific permissions related to workspaces in the project
+type WorkspaceAccessPermissions struct {
+	WorkspaceRunsPermission          WorkspaceRunsPermissionType          `jsonapi:"attr,runs"`
+	WorkspaceSentinelMocksPermission WorkspaceSentinelMocksPermissionType `jsonapi:"attr,sentinel-mocks"`
+	WorkspaceStateVersionsPermission WorkspaceStateVersionsPermissionType `jsonapi:"attr,state-versions"`
+	WorkspaceVariablesPermission     WorkspaceVariablesPermissionType     `jsonapi:"attr,variables"`
+	WorkspaceCreatePermission        bool                                 `jsonapi:"attr,create"`
+	WorkspaceLockingPermission       bool                                 `jsonapi:"attr,locking"`
+	WorkspaceMovePermission          bool                                 `jsonapi:"attr,move"`
+	WorkspaceDeletePermission        bool                                 `jsonapi:"attr,delete"`
+	WorkspaceRunTasksPermission      bool                                 `jsonapi:"attr,run-tasks"`
+}
```

Comment:
I think to find this struct under `team_project_access.go` would be slightly odd. Seems like something that should be defined under `workspace.go` or if only pertinent to team projects be prefixed with `TeamProjects` -- which would make the name absurdly long so to keep things. I prefer the former to circumvent a long name.

Reply:
This strucutre aligns with the API design of this endpoint. These perms also functionally cannot be part of workspace.go, since those are only for individual workspaces.

On the TeamProject level, there are two groups of permissions: one that only affects the project itself and those that are applied to all of the workspaces in the project. We thought that TeamProjectAccessWorkspaceAccessPermissions would be too long, but based on the nested struct of TeamProjectAccess would be clear enough.

Reply:
@sebasslash Thanks!
These really are scoped to team project access only (this current implementation is very close to the situation in the TeamAccess which doesn't prefix or hoist the struct defs into the workspace file).

Would it be better in this case to go with the longer names so that we can keep locality the same as it is with TeamAccess?

Reply:
Absolutely. I also think we can drop the second `Access`: `TeamProjectAccessWorkspacePermissions{}` and not lose any meaning here. Again these names can turn out to be quite long and difficult to remember, but the general rule would be to scope the name of the struct to the resource it belongs to. 

But yeah, `WorkspaceAccessPermissions` leads me to believe this is a struct that encapsulates all workspace permissions regardless of what resource I'm enabling those permissions for. So it'd might be best to avoid the confusion if these permissions are unique to team project access.

Reply:
Thanks @sebasslash! I will work on getting it all renamed by EOD today